### ISSUE TITLE
How would you go about running tests in nw?

### ISSUE THREAD

**Issue by sedouard** [brzpegasus/ember-cli-node-webkit#12]
I've done a few things to try to get our test suite running in on nwjs but it seems after a few hurdles (adding a testem loader for nwjs, removing the `'<base href="/"/>` ember cli throws in your index.html, copying a package.json to the `dist/tests` directory through broccoli) but  there's some weird behavior when it comes to loading modules in that we can't ever seem to load the `ember-data` module.

I can add some more details by sharing a fork but I was just curious from a discussion standpoint - how would you go about implementing this? Am I on the right track here?

**Comment by brzpegasus** [comment]
The testing story in NW.js is fairly lacking. At the moment, NW.js recommends using Selenium and [chromedriver](https://github.com/nwjs/nw.js/wiki/chromedriver) for automated functional tests. Of course, while those kinds of tests have their own benefits, they are slower to run, and you would still need to supplement with a bunch of unit tests.

I'm happy to work with you on trying to load the `tests/index.html` in NW.js, but I haven't personally gone that route yet as I'm not entirely convinced it is the right approach.

You would end up running a NW.js app to drive the tests, then load your Ember app within it, which isn't quite the same as spawning the actual application in an isolated NW.js environment for testing, and having the ability to reset or destroy/restart the app between tests.

What my team and I have done in a recent project was to structure the app in such a way that any functionality that depended on `nw.gui` or Node modules would be highly decoupled from the rest of the Ember app. This may not always be feasible but in our case, it allowed us to run the majority of our Ember tests in the browser (http://localhost:4200/tests), or with phantomjs (via `ember test`). If the app interfaced with Node or NW.js-specific components, those would get mocked or stubbed out. We tested the logic for our custom Node modules separately using good old Node-style tests (not in the browser), and wired up the `npm test` command to run both.

The [Markdown Editor demo app](https://github.com/brzpegasus/ember-nw-markdown) was created in a similar fashion. It runs with or without NW.js (`http://localhost:4200`), though of course, without NW.js, you get a degraded experience as there is no application menu or access to the Node file system API to open/save documents. That separation made both debugging and testing a lot more manageable.

I realize it may be a contrived example, and I'm definitely curious to hear from you or anyone else about their testing experience as there's definitely room here to make things better.

**Comment by sedouard** [comment]
I was thinking the same thing as you. Test the node portion separately and mock it out on the other side with the front-end web js. However my thought was more-so for TDD. I wanted the ability to quickly see if my stuff just 'worked' and it wasn't easy to test custom adapters and models which had node.js code mixed in. Also in the future it's not impossible that we may do more on the node.js side and I can imagine that running integration tests between vanilla js and node.js could be valuable.

I also was able to use `xvfb` on travis CI to run `nw` and make it appear to be 'headless'. With that and a custom Qunit logger we can output our unit test output in the console and even return a pass/fail to travis.

To give you some more context, we're working on a [cross platform storage explorer](https://github.com/azure-storage/xplat) client for Microsoft Azure and we're building it based on your template (this is awesome btw! :+1: ). To do this we're reusing the [azure-storage](https://www.npmjs.com/package/azure-storage) node module which is really intended to be used on the server-side. (Azure storage is similar to Amazon S3).

In our case we've implemented custom adapters in lieu of the ember REST adapter as well as models that are backed by the `azure-storage` model in order to manipulate blobs and containers and I really wanted our tests to be ran in nw.js so that going forward I could write more functionally into these adapters and models and easily test them using qunit. We could decide to mock out the node.js later if wanted.

Anyways here's [a fork](https://github.com/sedouard/xplat/tree/run_tests_in_nw) which attaches `ember test` to nw.js and runs our tests. The output shows that we can't load modules:

```
beforeEach failed on it should exist: Cannot find module 'azureexplorer/models/blob'
                   Error: Cannot find module 'azureexplorer/models/blob'\n    at Function.Module._resolveFilename (module.js:327:15)\n    at Function.Module._load (module.js:264:25)\n    at Module.require (module.js:356:17)\n    at require (module.js:375:17)\n    at window.require (\u003Canonymous\u003E:4:17)\n    at window.require (file:///Users/steve/Documents/xplat-sedouard/dist/assets/vendor.js:37:30)\n    at resolveOther (file:///Users/steve/Documents/xplat-sedouard/dist/assets/vendor.js:58910:20)\n    at superWrapper (file:///Users/steve/Documents/xplat-sedouard/dist/assets/vendor.js:27030:20)\n    at exports.default.EmberObject.default.extend.resolve (file:///Users/steve/Documents/xplat-sedouard/dist/assets/vendor.js:14349:35)
```

On the surface - it looks like we just tried to load a AMD module with node.js. What's actually happening is that `vendor/node-webkit/shim.js` is failing as it tries to load the AMD module `ember-data` because `DS is not defined` when it tires to pull it out of the built `vendor.js`. This causes a bunch of other module load failures and each module load failure falls back to trying load a node.js module which obviously doesn't work either. (You can see it if you remove the `catch` statement in `shim.js`.

Anyways - I thought I was sort of close to having some sort of prototype running unit tests in nw but I'm sort of stuck there.

Honestly at least as far as we are in this project, we'll probably end up taking your approach but I do think it would be sweet to be able to run tests in nw.js from a command line and get the results right there -- it would seem useful for a lot of other things, too.

**Comment by brzpegasus** [comment]
I just pulled down your fork on my machine. I think we can make this work. The error you're getting is actually because in `vendor/reload.js`, it is trying to watch for file changes in `dist`, which doesn't exist relative to `dist/tests/package.json`. It was a pretty naive thing for me to assume there would be a `dist` folder and I can change it up to either make the watch directory path configurable, or provide an option to opt out of the reload, or both.

For now, you can wrap the watch in a `fs.existsSync('./dist')`; something like this:

```javascript
// vendor/reload.js
(function() {
  if (!window.nwDispatcher) return;

  // Reload the page when anything in `dist` changes
  var fs = window.requireNode('fs');
  var watchDir = './dist';

  if (fs.existsSync(watchDir)) {
    fs.watch(watchDir, function() {
      window.location.reload();
    });
  }
})();
```

Once you do that, you should get most tests passing:

```

 TEST SUMMARY:
 PASSED: 50

 FAILED: 1
 TOTAL: 51
 Runtime: 356ms
```

The only failing test can easily be fixed. It's just not recognizing the global `QUnit` variable and such in `qunit-logger.js`.

**Comment by sedouard** [comment]
This is awesome! Thanks for finding that, I must have overlooked it a few times over :-).

Do you think this would be a good thing to PR into this repo? I think a couple things need to be tweaked like making the test output TAP compliant for tested to consume. The way I pull in the nw.js binary should also be examined.

Another hurdle really is that due to some behavior in ember cli the baseUrl for ember qunit is not compatible with running it in nw.js.

I'm happy to make this into a PR if you think it'll be good to have in here.

Thanks again!

**Comment by brzpegasus** [comment]
I would definitely love to have some kind of testing support in the addon, and you are more than welcome to contribute! I do like what you have going on at the moment and would be very interested to know how that ends up working for you in practice. For the nw.js binary, we may be able to just require it from NPM, instead of downloading it the way you do in `nw-runner.js` (by default, the addon already installs it into the project as an NPM dependency). Let me play with this a bit more, and if you open a PR, we could move the discussion there too.

**Comment by sedouard** [comment]
Yeah I actually just made [this change](https://github.com/sedouard/xplat/commit/9bf345a2ee059b0e5ed5e13260db5a0d9b43f0e3) in our project about an hour ago.

It runs in travis, I've added the TAP output from nw.js and with nwjs-runner massaging the output it allows testem to parse it properly. 
https://travis-ci.org/azure-storage/xplat/builds/57708297

Here's a failing example:
https://travis-ci.org/azure-storage/xplat/builds/57706793

I'll make a WIP pull req and we can chat there. I'd like to make this something optional for anyone using this plugin.

**Comment by brzpegasus** [comment]
:+1:

**Comment by brzpegasus** [comment]
@sedouard Going to close this out.

Just a quick update, I've delayed on releasing the new test command, but I'll hopefully get it out very soon. I wanted to iron out a few miscellaneous things to make the functionality as complete and usable as it can be before inviting others to use it. I've opened #23 to track some of those remaining items.

Thanks again for getting this going in the first place and for the PR!
